### PR TITLE
cmake flatcc build issue

### PR DESCRIPTION
**Pull request by haowhsu-quic**
Identical change to https://github.com/pytorch/executorch/pull/4312.

### CODE REVIEW

**Comment by cccclai** [comment]
Looks like the coreml delegate test failure is legit, [see log](https://productionresultssa6.blob.core.windows.net/actions-results/0a87402b-1402-476f-bcd9-45301fa77d0e/workflow-job-run-2ddd140f-a62a-5497-9cb6-e6aadb3ddda4/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-08-05T18%3A03%3A47Z&sig=G0m38QFBVgq9k1N1kYaN16Q%2F4Y%2BH7jxmVGt%2FfIacsIw%3D&ske=2024-08-06T05%3A39%3A48Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-08-05T17%3A39%3A48Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-05-04&sp=r&spr=https&sr=b&st=2024-08-05T17%3A53%3A42Z&sv=2024-05-04)

```

2024-08-04T00:04:12.0402250Z Ld /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/InstallableProducts/usr/local/bin/coreml_executor_runner normal (in target 'coreml_executor_runner' from project 'coreml_executor_runner')
--
  | 2024-08-04T00:04:12.0403920Z     cd /Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/executor_runner
  | 2024-08-04T00:04:12.0416230Z     /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -Xlinker -reproducible -target arm64-apple-macos13.0 -isysroot /Applications/Xcode_15.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -O0 -L/Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/Intermediates.noindex/EagerLinkingTBDs/Debug -L/Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/xcode-build/Debug -L/Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/executor_runner/libraries -L/Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/executor_runner/libraries -F/Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/Intermediates.noindex/EagerLinkingTBDs/Debug -F/Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/xcode-build/Debug -filelist /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/Intermediates.noindex/coreml_executor_runner.build/Debug/coreml_executor_runner.build/Objects-normal/arm64/coreml_executor_runner.LinkFileList -Xlinker -object_path_lto -Xlinker /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/Intermediates.noindex/coreml_executor_runner.build/Debug/coreml_executor_runner.build/Objects-normal/arm64/coreml_executor_runner_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -fobjc-arc -fobjc-link-runtime -all_load -letdump -lexecutorch_no_prim_ops -lflatccrt -lcoremldelegate -framework Accelerate -lprotobuf-lite -lportable_ops_lib -framework CoreML -lportable_kernels -lsqlite3 -lexecutorch -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/Intermediates.noindex/coreml_executor_runner.build/Debug/coreml_executor_runner.build/Objects-normal/arm64/coreml_executor_runner_dependency_info.dat -o /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/InstallableProducts/usr/local/bin/coreml_executor_runner
  | 2024-08-04T00:04:12.0427680Z ld: library 'flatccrt' not found
  | 2024-08-04T00:04:12.0428290Z clang: error: linker command failed with exit code 1 (use -v to see invocation)
  | 2024-08-04T00:04:12.0428630Z
  | 2024-08-04T00:04:12.0428690Z ** BUILD FAILED **
  | 2024-08-04T00:04:12.0428810Z
  | 2024-08-04T00:04:12.0428980Z
  | 2024-08-04T00:04:12.0429140Z The following build commands failed:
  | 2024-08-04T00:04:12.0430710Z 	Ld /Users/runner/Library/Developer/Xcode/DerivedData/coreml_executor_runner-bkoiqshilmuvptcxjvpygyufxpnf/Build/Intermediates.noindex/InstallIntermediates/coreml_executor_runner-macosx/InstallableProducts/usr/local/bin/coreml_executor_runner normal (in target 'coreml_executor_runner' from project 'coreml_executor_runner')
  | 2024-08-04T00:04:12.0432030Z (1 failure)
  | 2024-08-04T00:04:12.0432850Z cp: /Users/runner/work/executorch/executorch/pytorch/executorch/examples/apple/coreml/xcode-build/DEBUG/coreml_executor_runner: No such file or directory
  | 2024-08-04T00:04:12.0433630Z ExecuTorch: Failed to build executor.
  | 2024-08-04T00:04:12.0434210Z ERROR conda.cli.main_run:execute(47): `conda run bash backends/apple/coreml/scripts/build_all.sh` failed. (See above for error)
  | 2024-08-04T00:04:12.0434910Z     main()
  | 2024-08-04T00:04:12.0435510Z   File "/Users/runner/work/executorch/executorch/test-infra/.github/scripts/run_with_env_secrets.py", line 61, in main
  | 2024-08-04T00:04:12.0436520Z     run_cmd_or_die(f"bash { os.environ.get('RUNNER_TEMP', '') }/exec_script")
  | 2024-08-04T00:04:12.0437410Z   File "/Users/runner/work/executorch/executorch/test-infra/.github/scripts/run_with_env_secrets.py", line 39, in run_cmd_or_die
  | 2024-08-04T00:04:12.0438160Z     raise RuntimeError(f"Command {cmd} failed with exit code {exit_code}")
  | 2024-08-04T00:04:12.0438770Z RuntimeError: Command bash /Users/runner/work/_temp/exec_script failed with exit code 1
  | 2024-08-04T00:04:12.0489710Z ##[error]Process completed with exit code 1.
  | 2024-08-04T00:04:12.0599850Z ##[group]Run pmeier/pytest-results-action@v0.3.0

```

**Comment by cccclai** [comment]
Do you have a mac machine to repro?

**Comment by haowhsu-quic** [comment]
I wonder if this job is making debug build? Since coreml explicitly requires [release library](https://github.com/pytorch/executorch/blob/ec4f576872fcb834e198cde9bec5598ab3a836c6/examples/apple/coreml/scripts/build_executor_runner.sh#L72C46-L72C103).
We'll try to get a mac machine for further investigation, thank you.

**Comment by haowhsu-quic** [comment]
Could you help make this [file](https://github.com/pytorch/executorch/blob/f52d8ab6ac86eacc18eb3f9503221b149eaefc1d/examples/apple/coreml/scripts/build_executor_runner.sh#L31) into release build (or at least -DCMAKE_BUILD_TYPE=RelWithDebInfo)?

I think this might be the root cause in following log coming from [here](https://github.com/pytorch/executorch/blob/ec4f576872fcb834e198cde9bec5598ab3a836c6/examples/apple/coreml/scripts/build_executor_runner.sh#L72C46-L72C103): 
image

**Comment by chiwwang** [comment]
It's not clear why the PR can affect Mac build... mmmm... maybe it's because the removal of the line `add_dependencies(etdump_schema flatcc_cli)`?
Finding a Mac to try....

Another way to resolve the race-condition between the target and the host environment: https://github.com/pytorch/executorch/pull/4559, which might have smaller changes in CMakeList.
But still, need to find a Mac to try....

**Comment by cccclai** [comment]
Catching up....which part makes you feel like release build is the issue here?

**Comment by haowhsu-quic** [comment]
Hi, I add `-DCMAKE_BUILD_TYPE=Release` to this [script](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L30). Coreml explicitly requests [release version of flatccrt](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L72), but the trunk job has no way to set this configuration (backends/apple/scripts/build_all.sh).
I change it with latest commit, all jobs are working properly now.

**Comment by chiwwang** [comment]
> Hi, I add `-DCMAKE_BUILD_TYPE=Release` to this [script](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L30). Coreml explicitly requests [release version of flatccrt](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L72), but the trunk job has no way to set this configuration (backends/apple/scripts/build_all.sh). I change it with latest commit, all jobs are working properly now.

Just trying to clarify.
Is the path `$EXECUTORCH_ROOT_PATH/third-party/flatcc/lib/libflatccrt.a` dedicated for release-build? 
I guess if flatcc is not release-built, the `libflatccrt.a` would be in another place. Is this part making you feel like release build is the issue?

**Comment by haowhsu-quic** [comment]
> > Hi, I add `-DCMAKE_BUILD_TYPE=Release` to this [script](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L30). Coreml explicitly requests [release version of flatccrt](https://github.com/pytorch/executorch/blob/05a7d52510577565f0c773546f4caa5cd43a876f/examples/apple/coreml/scripts/build_executor_runner.sh#L72), but the trunk job has no way to set this configuration (backends/apple/scripts/build_all.sh). I change it with latest commit, all jobs are working properly now.
> 
> Just trying to clarify. Is the path `$EXECUTORCH_ROOT_PATH/third-party/flatcc/lib/libflatccrt.a` dedicated for release-build? I guess if flatcc is not release-built, the `libflatccrt.a` would be in another place. Is this part making you feel like release build is the issue?

Yes, for debug-build library name will be `libflatccrt_d.a`. We also reproduced the error internally.

**Comment by cccclai** [comment]
cc: @Olivia-liu @tarun292 @dbort it really should be our job to make sure flatcc can be built correctly, and thanks folks from Qualcomm helping this.

File: examples/apple/coreml/scripts/build_executor_runner.sh [hunk 1]
```diff
@@ -29,6 +29,7 @@ rm -rf "$CMAKE_BUILD_DIR_PATH"
 # Build executorch
 echo "ExecuTorch: Building executorch"
 cmake "$EXECUTORCH_ROOT_PATH" -B"$CMAKE_BUILD_DIR_PATH" \
+-DCMAKE_BUILD_TYPE=Release \
```
Comment (cccclai) [inline examples/apple/coreml/scripts/build_executor_runner.sh:32]: @cymbalrush for the change in the coreml script

**Review by cccclai — APPROVED** [review]
lgtm, will wait a bit for the response.

**Comment by Olivia-liu** [comment]
> cc: @Olivia-liu @tarun292 @dbort it really should be our job to make sure flatcc can be built correctly, and thanks folks from Qualcomm helping this.

Agreed. I will make sure to allocate time to improve the build process for flatcc.

**Pull request merged**